### DISCUSSION TITLE
Supporting multiple windows on iOS

### DISCUSSION THREAD

**Discussion by fabwu** [organicmaps/organicmaps#6792] [Ideas]
Hi,

I'm not quite happy with the CarPlay experience of OrganicMaps and for me the two most pressing issues are #763 (app is unusable while connected to CarPlay) and #6184 (CarPlay dashboard integration). Both of these issues need proper support for multiple windows or the scene-based app lifecycle introduced with iOS13. I've done some research and found these resources:

[Session 212 “Introducing Multiple Windows on iPad”](https://developer.apple.com/videos/play/wwdc2019/212) 
[Session 258 “Architecting Your App for Multiple Windows”](https://developer.apple.com/videos/play/wwdc2019/258)
[Session 246 “Window Management in Your Multitasking App”](https://developer.apple.com/videos/play/wwdc2019/246)
[Session 242 “Targeting Content with Multiple Windows”](https://developer.apple.com/videos/play/wwdc2019/242)
https://adrian.schoenig.me/blog/2019/06/13/multi-window-support-on-ipad/
https://www.kairadiagne.com/2019/11/18/adopting-the-new-scene-delegate

After watching the videos, I started a humble first attempt, which you can find [here](https://github.com/fabwu/organicmaps/tree/ios_multi_window). Now the problem is I've just started iOS development and I don't really know what I'm doing. Anyway, the app starts, but creating multiple windows on iPad and CarPlay doesn't work. I think the biggest obstacle is the shared MapViewController, which now exists multiple times.

I just wanted to get some feedback or your opinion before I continue. It would also be great if a more experienced iOS developer could give me some hints or advice.

I'm happy to hear from you!

Fabian

**Comment by biodranik** [comment]
Hi and thanks for the help on these issues! The main problem is with the underlying Metal/OpenGL rendering code that is sharing now the same screen context between the CarPlay and iOS devices. Obviously, screens are different and should not be shared. There is a similar issue with Android Auto. A separate issue can be created (if it's not there already) to refactor the Drape rendering engine and support several graphics contexts (screens) for rendering. 

@AndrewShkrob @vng do we already have that issue? Can you please create it and link it with related iOS/Android issues?

**Reply by fabwu** [reply]
Thanks for your quick reply. I‘ve found this PR #4080 but I‘m not sure if I can already use this code?

I was wondering how CarPlay works at the moment. Is the context of from the app screen passed to the CarPlay screen and then rescaled?
